File: db/index_building/py/buildIndex.py
import pymongo 
import nltk 
from nltk.corpus import stopwords
nltk.download('stopwords')
import re
from nltk.corpus import stopwords
nltk.download('stopwords')
from nltk.stem.porter import *
import pickle
import gc
from dotenv import load_dotenv
from os.path import join, dirname
dotenv_path = join(dirname(__file__), '.env')
load_dotenv(dotenv_path)
import os

def build_index(): # each term will have 200 docs max
    stopSet = set(stopwords.words('english'))
    stemmer = PorterStemmer()
    DB_PASS= os.environ.get("DB_PASS")
    DB_USER= os.environ.get("DB_USER")
    DB_NAME= os.environ.get("DB_NAME")
    DB_HOST= os.environ.get("DB_HOST")
    PORT = os.environ.get("PORT")
    client = pymongo.MongoClient(f'mongodb://{DB_USER}:{DB_PASS}@{DB_HOST}:{PORT}') 
    db = client["TTDS"]
    quotesCollec = db["quotes"]
    loadSize = 1000
    currIndexPortion = dict()
    iterations = 0
    files_counter = 0
    
    for r in range(0, 41000):
        iterations += 1
        lowerLimit = r * loadSize
        upperLimit = lowerLimit + loadSize

        # Load 1000 quote objects at a time in memory
        quotes = quotesCollec.find({"_id" : {"$lt" : upperLimit, "$gte" : lowerLimit}})
        
        # Populate currIndexPortion with the current quotes batch
        for q in quotes:
            q_id = q['_id']
            b_id = q['book_id']
            regexForPos = re.compile("[\s.,;'\"\(\)\[\]]")
            termsSplit = regexForPos.split(q['quote'].lower())
            terms = [stemmer.stem(token.lower()) for token in re.findall(r'\w+', q['quote']) if not token.lower() in stopSet]

            for term in set(terms):
                pos = [i for i, split in enumerate(termsSplit) if term in split]
                currIndexPortion[term] = currIndexPortion.get(term, {'term' : term, 'term_freq' : 0, 'books' : {}})
                currIndexPortion[term]['term_freq'] += 1
                currIndexPortion[term]['books'][b_id] = currIndexPortion[term]['books'].get(b_id, {'_id': b_id, 'term_freq_in_book': 0, 'quotes': []})
                currIndexPortion[term]['books'][b_id]['term_freq_in_book'] += 1
                currIndexPortion[term]['books'][b_id]['quotes'].append({'_id' : q_id, 'len' : len(re.findall(r'\w+', q['quote'])), 'pos': pos})

        # Write current batch to disk once every 205 iterations: will output 200 portions of the index
        if iterations == 205:
            iterations = 0
            for value in currIndexPortion.values():
                value['books'] = list(value['books'].values())
            # Write to disk in the Mongo insert_many format
            with open("/root/index/" + str(files_counter) + '.p', 'wb') as handle:
                pickle.dump(list(currIndexPortion.values()), handle, protocol=pickle.HIGHEST_PROTOCOL)
                files_counter += 1
            currIndexPortion.clear()
            gc.collect()

    for value in currIndexPortion.values():
        value['books'] = list(value['books'].values())
    # Write to disk in the Mongo insert_many format
    with open("/root/index/" + str(files_counter) + '.p', 'wb') as handle:
        pickle.dump(list(currIndexPortion.values()), handle, protocol=pickle.HIGHEST_PROTOCOL)
    currIndexPortion.clear()
    gc.collect()

    client.close()

       
build_index()
     

File: db/index_building/js/buildIndex.js
const fs = require('fs').promises;
const MongoClient = require('mongodb').MongoClient;
const url = "mongodb://localhost";
const stemmer = require('porter-stemmer').stemmer;

let buildIndex = async () => {
    let text = await fs.readFile('./nltk_stop_words.txt', 'utf8'); 
    let stopWords = new Set(text.split("\n"));
    let client = await MongoClient.connect(url, { useUnifiedTopology : true});
    

    try {
        let quotesCollec = client.db("TTDS").collection("quotes");
        let numOfQuotes = await client.db("TTDS").collection("quotes").countDocuments();
        let loadSize = 1000
        let iterations = 0 
        let current_file = 0
        let dict = {}

        // Load quotes in 1,000 at a time
        for (let r = 0; r < 57000; r++) {
            let lowerLimit = r * loadSize;
            let upperLimit = lowerLimit + loadSize;
            let quotes = await quotesCollec.find({"_id" : {"$lt" : upperLimit, "$gte" : lowerLimit}}).toArray();
            
            for (i in quotes) {
                let quote = quotes[i];
                let q_id = quote['_id'];
                let b_id = quote['book_id'];
                let terms = await preprocess(quote['quote'], stopWords);
                let uniqueTerms = Array.from(new Set(terms));

                for (j in uniqueTerms) {
                    let term = uniqueTerms[j];
                    let pos = getPositionsOfTerm(term, terms);
                    dict[term] = dict[term] || {'term' : term, 'term_freq' : 0, 'books' : {}};
                    dict[term]['term_freq'] += 1;
                    dict[term]['books'][b_id] = dict[term]['books'][b_id] || {'_id': b_id, 'term_freq_in_book': 0, 'quotes': []};
                    dict[term]['books'][b_id]['term_freq_in_book'] += 1;
                    dict[term]['books'][b_id]['quotes'].push({'_id' : q_id, 'len' : terms.length, 'pos': pos})
                }
            }
            
            if (iterations == 284) {
                iterations = 0;
                let keys = Object.keys(dict);
                for (k in keys) {
                    let term = keys[k];
                    dict[term]['books'] = Object.values(dict[term]['books']);
                }

                let file = await fs.writeFile(`/root/index/${current_file}.json`, JSON.stringify(Object.values(dict)), 'utf-8');
                current_file++;
                dict = null;
            }
        
            console.log("Processed ", upperLimit, "quotes out of ", numOfQuotes);
        }   

        let keys = Object.keys(dict);
        for (k in keys) {
            let term = keys[k];
            dict[term]['books'] = Object.values(dict[term]['books']);
        }

        let file = await fs.writeFile(`/root/index/${current_file}.json`, JSON.stringify(Object.values(dict)), 'utf-8');
        dict = null;

    } finally {
        console.log("Done. Press Ctrl + C to exit program.")
    }
};


let getPositionsOfTerm = (term, terms) => {
    let result = [];
    for (let i = 0; i < terms.length; i++) {
        if (term == terms[i]) {
            result.push(i);
        }
    }
    return result;
}

let preprocess = async (quote, stopWords) => {
    let final = [];
    let tokens = quote.match(/\w+/gm);
    for (i in tokens) {
        let token = tokens[i].toLowerCase();
        if (!stopWords.has(token)) {
            final.push(stemmer(token));
        }
    }
    return final;
}

let run = async () => {
    let start = new Date().getTime();
    try {
        await buildIndex();
    } catch(err) {
        console.log("Error in buildIndex(): ", err);
    }
    
    let end = new Date().getTime();
    let seconds = (end - start) / 1000;
    let minutes = seconds / 60;
    console.log("Finished building index in ", seconds , " seconds or ", minutes, " minutes.");
}

run();
